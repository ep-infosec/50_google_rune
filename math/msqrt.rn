//  Copyright 2021 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Find a square root modulo a prime number p.  Return 0 if it does not exist.
// The other root is the negation of the root returned.
export func msqrt(a, p) {
  q = a mod p
  if p == <p>2 {
    return a
  }
  // First check that the square root exists.
  if a**((p-<p>1)/<p>2) != 1 mod p {
    return <p>0
  }
  if !<u2>p == 3u2{
    // p == 3 mod 4
    return a**((p+<p>1)/<p>4) mod p
  }
  if !<u3>p == 5u3 {
    // p == 5 mod 8
    v = (2*a)**((p-<p>5)/<p>8) mod p
    im = 2*a*v**2 mod p
    return a*v*(im-1) mod p
  }
  // Use Shanks algorithm.  Find 2**e*q == p - 1.
  s = 1
  q = (p-<p>1)/<p>2
  while !<u1>q == 0u1 {
    q /= <p>2
    s += 1
  }
  // Find a non-square mod p.
  z = 2
  while z**((p-<p>1)/<p>2) == 1 mod p {
    z += 1
  }
  m = s
  c = z**q mod p
  t = a**q mod p
  r = a**((q+<p>1)/<p>2) mod p
  while true {
    if t == <p>0 {
      return <p>0
    }
    if t == <p>1 {
      return r
    }
    i = 1
    ts = t**2 mod p
    while ts != <p>1 mod p {
      ts = ts**2 mod p
      i += 1
    }
    b = c**(2**(m-i-1)) mod p
    m = i
    c = b**2 mod p
    t = t*c mod p
    r = r*b mod p
  }
  return <p>0  // Dummy return
}

unittest msqrtTest {
  for i in range(5) {
    p = findRandPrime()
    println "Found prime ", p
    for j in range(10) {
      a = reveal(rand100 % p)
      r = msqrt(a, p)
      if r == <p>0 {
        println "Not a square: ", a
      } else {
        println "a = ", a, ", r = ", r, ", r**2 = ", r**2 mod p
        if r != <p>0 {
          if r**2 != a mod p {
            throw "Error: %u100**2 != %u100 mod %u100\n" % (r, a, p)
          }
        }
      }
    }
  }
  println "Passed"

  func findRandPrime() {
    do {
      p = reveal(rand100)
    } while !isPrime(p)
    return p
  }

  func isPrime(p) {
    if p < <p>2 {
      return false
    }
    for i in range(32) {
      w = reveal(rand100) % (p - <p>1) + <p>1
      l = w**(p-<p>1) mod p
      if l != <p>1 {
        return false
      }
    }
    return true
  }
}
