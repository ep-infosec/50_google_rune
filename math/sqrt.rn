//  Copyright 2021 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

func sqrt(squared: f64, tolerance: f64 = 0.00000000000003) {
  guess = squared / .115f64
  nextGuess = 0.5 * (guess + (squared / guess))
  while abs(nextGuess - guess) > tolerance {
    guess = nextGuess
    nextGuess = 0.5 * (guess + (squared / guess))
  }

  return nextGuess
}
