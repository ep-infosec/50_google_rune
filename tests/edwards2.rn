//  Copyright 2021 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import math

class EdwardsCurve(self, <m>, d: typeof(m), gy: typeof(m)) {
  self.m = m  // Modulus
  self.d = d  // Edwards curve d parameter
  self.gy = gy
  self.gx = self.computeX(gy)

  // Verify (gx, gy) is on the curve: -x**2 + y**2 = 1 + dx**2y**2:
  func pointOnCurve(self, P) -> bool {
    x = P.x; y = P.y
    m = self.m; d = self.d
    return -x**2 + y**2 == 1 + d*x**2*y**2 mod m
  }

  // Return the positive X coordinate given Y.
  func computeX(self, y) {
    root = math.msqrt((y**2 - 1)/(self.d*y**2 + 1) mod self.m, self.m)
    return min(root, self.m - root)
  }
}

// A point on the curve.
class Point(self, <curve>: EdwardsCurve, x: typeof(curve.m), y: typeof(curve.m)) {
  self.curve = curve
  self.x = x
  self.y = y
  if !curve.pointOnCurve(self) {
    throw "The point is not on the curve"
  }

  operator + (a: Point, b: Point) -> Point {
    curve = a.curve
    if curve != b.curve {
      throw "Points are on different curves"
    }
    d = curve.d
    m = curve.m
    x = (a.x*b.y + a.y*b.x) / (<d>1 + d*a.x*a.y*b.x*b.y) mod m
    y = (a.x*b.x + a.y*b.y) / (<d>1 - d*a.x*a.y*b.x*b.y) mod m
    return Point(curve, x, y)
  }

  // Only scalar multiplication is defined for elliptic curve points.
  operator * (k: Uint, P: Point) {
    R: typeof(P) = Point(P.curve, <P.x>0, <P.y>1)
    M: typeof(P) = P  // M is 2**n*P
    f = k
    while f != <f>0 {
      if f & <f>1 != <f>0 {
        R += M
      }
      M += M
      f >>= 1
    }
    return R
  }
}

unittest edwardsTest {
  // Curve 25519 parameters.
  m = <u255>(2u256**255 - 19)
  d = -121665/121666 mod m
  // Curve 25519 generator's Y coordinate.
  gy = 4 / 5 mod m
  println "gy * 5 = ", gy * 5 mod m
  curve = EdwardsCurve(m, d, gy)
  G = Point(curve, curve.gx, curve.gy) // Generateor base point
  println "gx = ", G.x
  println "gy = ", G.y
  a = 5678u255  // Alice's private key
  A = a * G  // Alice's public key
  b = 1234u255  // Bob's private key
  B = b * G  // Bob's public key
  aliceSecret = a * B
  bobSecret = b * A
  println "Alice's secret = %x" % aliceSecret.x
  println "Bob's secret = %x" % bobSecret.x
  if aliceSecret.x != bobSecret.x {
    throw "Failed"
  }
  println "Passed"
}
