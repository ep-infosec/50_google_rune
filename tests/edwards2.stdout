gy * 5 = 4
gx = 15112221349535400772501151409588531511454012693041857206046113283949847762202
gy = 46316835694926478169428394003475163141307993866256225615783033603165251855960
Alice's secret = 3c022502fab4dc16468bf32771421fd42bbd761d0cb884f98221cf85326a56d7
Bob's secret = 3c022502fab4dc16468bf32771421fd42bbd761d0cb884f98221cf85326a56d7
Passed
